pkg/daemon/cluster.go

```diff
@@ -32,5 +36,81 @@ type Cluster struct {
 	ConfDirHostPath string
 }
 
+type DeploymentConfig struct {
+	Name           string
+	NodeName       string
+	Type           string
+	Labels         map[string]string
+	Volumes        []v1.Volume
+	Containers     []v1.Container
+	InitContainers []v1.Container
+	*Cluster
+}
+
+const (
+	SnapshotcloneAppName = "curve-snapshotclone"
+	GrafanaAppName       = "curve-grafana"
+	EtcdAppName          = "curve-etcd"
+)
+
 func (c *Cluster) GetUUID() string { return c.UUID }
 func (c *Cluster) GetKind() string { return c.Kind }
+
+func MakeDeployment(c DeploymentConfig) (*apps.Deployment, error) {
+	spec := v1.PodSpec{
+		Containers:    c.Containers,
+		NodeName:      c.NodeName,
+		RestartPolicy: v1.RestartPolicyAlways,
+		HostNetwork:   true,
+		DNSPolicy:     v1.DNSClusterFirstWithHostNet,
+		Volumes:       c.Volumes,
+	}
+
+	switch c.Type {
+	case GrafanaAppName, SnapshotcloneAppName:
+		runAsUser := int64(0)
+		runAsNonRoot := false
+		spec.SecurityContext = &v1.PodSecurityContext{
+			RunAsUser:    &runAsUser,
+			RunAsNonRoot: &runAsNonRoot,
+		}
+	case EtcdAppName:
```

Comment:
if one pod has no initContainer, the initContiner is nil. not need to discuss on a case-by-case basis because we want a more common method, it should shielding details.

Reply:
done